doc/user_guide/psy_data.rst

```diff
@@ -0,0 +1,302 @@
+  declaration, or because memory is overwritten). A checksum can be
+  used to detect any changes to a read-only field. Again, this is 
+  a feature for the future and not available at this stage.
+
+The PsyData API should be general enough to allow these and other
+applications to be developed and used.
+
+API
+---
+The callbacks are inserted into the code by the PSyData transformation,
+see :ref:`psy_data_transformation` for details. The following example
+shows the code created by PSyclone::
+
+    USE psy_data_mod, ONLY: PSyDataType
+    TYPE(PSyDataType), save :: psy_data
+
+    CALL psy_data%PreStart("update_field_mod", "update_field_code", 1, 1)
+    CALL psy_data%PreDeclareVariable("a_fld", a_fld)
+    CALL psy_data%PreDeclareVariable("b_fld", b_fld)
+    CALL psy_data%PreEndDeclaration
+    CALL psy_data%ProvideVariable("a_fld", a_fld)
+    CALL psy_data%PreEnd
+
+    ! Begin of PSY-layer kernel call:
+    DO j=1,jstop+1
+      DO i=1,istop+1
+        CALL update_field_code(i, j, a_fld%data, b_fld%data)
+      END DO 
+    END DO 
+    ! End of PSY-layer kernel call
+
+    CALL psy_data%PostStart
+    CALL psy_data%ProvideVariable("b_fld", b_fld)
+    CALL psy_data%PostEnd
+
+The following code sequence will typically be created:
+
+#. A static variable of type ``PSyDataType`` is declared.
+#. ``PreStart`` is called to indicate that a new that a new instrumented code
+   region is started.
+#. ``PreDeclareVariable`` is called for each variable passed to the data API
+   either before or after the instrumented region.
+#. ``PreEndDeclaration`` is called to indicate the end of the variable
+   declarations.
+#. ``ProvideVariable`` is called for each variable to be written before the
+   instrumented region.
+#. ``PreEnd`` is called to signal the end of all PSyData activity before the
+   instrumented region.
+#. Then the actual instrumented code region is entered.
+#. After the instrumented region, a call to ``PostStart``
+   is added to indicate that all further data output occurs after the
+   instrumented region.
+
+#. For each variable to be written after the instrumented region a call to
+   ``ProvideVariable`` is added.
+#. A call to ``PostEnd`` is added once all variables have been written.
+
+.. note::
+    Depending on the options provided to the PSyData transformation
+    some of the calls might not be created. For example, for a performance
+    profiling library no variables will be declared or written.
+
+The following sections will describe the API in detail.
+
+.. _psy_data_type:
+
+``PSyDataType``
++++++++++++++++
+The library using the PSyData API must provide a user-defined data type
+called ``PSyDataType``. It is up to the application how this variable is
+used. PSyclone will declare the variables to be static, meaning that they
+can be used to accumulate data from call to call.
+
+``PreStart(this, module_name, kernel_name, num_pre_vars, num_post_vars)``
```

Comment:
I think this would be nicer if you used sphinx markup, e.g. `.. method::PreStart(this...)` but I'm not sure whether you can? Worth a quick look.

Reply:
Mostly agreed. It took me a while to figure out how to make it look nice ... I didn't know that you had to intend the description following the "method" heading. The only thing missing now is listing the API calls in the TOC, I couldn't declare them as chapter headings, nor figure out a way of adding entries to the TOC. But am happy the way it is.

Reply:
It looks great. Thanks.